ValidatedQueries: start of validated queries API

jalevin opened the pull request:
**What this PR does / why we need it**:

Adds a basic API for validated queries (queries that are checked against what's saved in the dashboard). For now, it only makes sure that the provided dashboard and panel actually exist.

Relates to https://github.com/grafana/grafana-enterprise-partnerships-team/issues/67

pianohacker committed:
WIP: start of tests

jalevin committed:
added 404 test which invalidates first test. need to figure out how in-depth we want to go on panelId validation

jalevin committed:
api endpoint now validates datasource exists and panel belongs to the dashboard

jalevin committed:
reworking to query dashboard instead of datasource .facepalm

jalevin committed:
swap datasource over to dashboard

leandro-deveikis commented on pkg/api/metrics.go:
```diff
@@ -40,6 +42,80 @@ func (hs *HTTPServer) QueryMetricsV2(c *models.ReqContext) response.Response {
+
+	if err := bus.Dispatch(c.Req.Context(), &query); err != nil {
+		return false
+		//if errors.Is(err, models.ErrDataSourceNotFound) {
```
Return same error as bus.Dispatch

leandro-deveikis commented on pkg/api/metrics.go:
```diff
@@ -40,6 +42,80 @@ func (hs *HTTPServer) QueryMetricsV2(c *models.ReqContext) response.Response {
+
+	if dashboardId == "" || panelId == "" {
+		// NOTE: Is this an appropriate status code?
+		return response.Error(http.StatusForbidden, "missing dashboard or panel ID", nil)
```
Return BadRequest (400)

pianohacker committed:
WIP: refactoring

leandro-deveikis committed:
Adding validation for feature flag ValidatedQueries

leandro-deveikis committed:
Changing error code to 409 when validated queries is disabled

pianohacker committed:
Fix response code when validated queries not enabled

pianohacker committed:
Finish tests

jalevin commented:
Hey @kminehart would you be able to give us a review?

kylebrandt commented on pkg/api/metrics.go:
```diff
@@ -40,6 +44,101 @@ func (hs *HTTPServer) QueryMetricsV2(c *models.ReqContext) response.Response {
+	// Dashboard has no properties. This would be weird, haven't written a test
+	// for this yet, but would most likely be a bug and return an error from
+	// bus.Dispatch
+	if query.Result == nil {
```
Do we recover from panics in calling code here? Even if unlikely seems like the behavior of panicing the process would undesirable.

jalevin replied:
We're not recovering from a panic here currently. Our thought process was that if this ever actually happens we have a serious problem and should shutdown. 

Happy to return an error and keep running if you think that makes more sense.

kylebrandt replied:
I think better to change it given that Grafana is still largely a monolith so recycle time can be a bit. And if there was a dashboard making a bunch of validate queries might take out multiple instances depending on the lb setup.

kylebrandt commented on pkg/api/metrics.go:
```diff
@@ -40,6 +44,101 @@ func (hs *HTTPServer) QueryMetricsV2(c *models.ReqContext) response.Response {
+
+	if err := bus.Dispatch(ctx, &query); err != nil {
+		return err
+		//if errors.Is(err, models.ErrDataSourceNotFound) {
```
commened out code should be removed.

kylebrandt commented on pkg/api/metrics.go:
```diff
@@ -40,6 +44,101 @@ func (hs *HTTPServer) QueryMetricsV2(c *models.ReqContext) response.Response {
+		Id: id,
+	}
+
+	if err := bus.Dispatch(ctx, &query); err != nil {
```
We are working on removing the bus. Can the SQL store be passed in instead (property on `hs *HTTPServer`) and its `GetDashboard` method be used instead?

kylebrandt commented on pkg/api/api.go:
```diff
@@ -381,6 +381,9 @@ func (hs *HTTPServer) registerRoutes() {
 		apiRoute.Post("/ds/query", authorize(reqSignedIn, ac.EvalPermission(ActionDatasourcesQuery)), routing.Wrap(hs.QueryMetricsV2))
 
+		// Validated query
+		apiRoute.Post("/dashboards/id/:dashboardId/panels/:panelId/query", authorize(reqSignedIn, ac.EvalPermission(ActionDatasourcesQuery)), routing.Wrap(hs.QueryMetricsFromDashboard))
```
I think we generally prefer going with UIDs over IDs, it looks like GetDashboard can work by uid, do we need to be using id?

jalevin replied:
We didn't know that. Happy to make the change!

jalevin committed:
remove comments

jalevin committed:
not sure if this is the approach we want to take for fixing up the mock

jalevin committed:
checkpoint. able to use out mock sql db

jalevin committed:
Refactoring references to sqlstore.SQLStore to sqlstore.store

This commit refactors test cases to use the interface sqlstore.Store
instead of sqlstore.SQLStore so that we can swap out _real_ sqlite
instances for mock instances.

jalevin committed:
checkpoint. refactored api to require orgId. moved parse params into own function

jalevin committed:
checkpoint

leandro-deveikis committed:
Adding testing on ParseDashboardQueryParam func

jalevin committed:
refactor checkDashboardAndPanel test

leandro-deveikis committed:
finishing testing for ParseDashboardQueryParams

jalevin committed:
test suite working. need to refactor common_test

jalevin committed:
full test suite working. plumbed in mock db

jalevin committed:
Merge branch 'main' into pianohacker/48-basic-validated-queries-api

jalevin committed:
Merge branch 'main' into pianohacker/48-basic-validated-queries-api

jalevin committed:
Merge branch 'pianohacker/48-basic-validated-queries-api' of github.com:grafana/grafana into pianohacker/48-basic-validated-queries-api

jalevin committed:
fix linting error

jalevin committed:
fix linting

jalevin committed:
linting...

ying-jeanne commented on pkg/api/common_test.go:
```diff
@@ -337,21 +338,36 @@ func setupHTTPServer(t *testing.T, useFakeAccessControl bool, enableAccessContro
 	return setupHTTPServerWithCfg(t, useFakeAccessControl, enableAccessControl, cfg)
 }
 
+func setupHTTPServerWithMockDb(t *testing.T, useFakeAccessControl bool, enableAccessControl bool) accessControlScenarioContext {
```
nice

ying-jeanne reviewed (approved):
LGTM
